File: beaglesightlibs/src/main/java/com/cross/beaglesightlibs/PositionCalculator.java
package com.cross.beaglesightlibs;

import java.text.DecimalFormat;
import java.util.List;

public abstract class PositionCalculator
{
    List<PositionPair> positions;
    private static final DecimalFormat hn = new DecimalFormat("#");
    private static final DecimalFormat df = new DecimalFormat("#.##");

	void setPositions(List<PositionPair> pos) {
		positions = pos;
	}

	public abstract float calcPosition(float distance);

    public static String getDisplayValue(float val, int numPlaces) {
        switch (numPlaces) {
            default:
            case 0:
                return hn.format(val);
            case 2:
                return df.format(val);

        }
    }

    //TODO: Horrifically slow and inaccurate. Replace with optimised version?
    public float getMaxPosition(float distStart, float distEnd)
    {
        float max = Float.MIN_VALUE;
        for (float i = distStart; i < distEnd; i++)
        {
            float val = this.calcPosition(i);
            if (val > max)
            {
                max = val;
            }
        }
        return max;
    }

    //TODO: Horrifically slow and inaccurate. Replace with optimised version?
    public float getMinPosition(float distStart, float distEnd)
    {
        float min = Float.MAX_VALUE;
        for (float i = distStart; i < distEnd; i++)
        {
            float val = this.calcPosition(i);
            if (val < min)
            {
                min = val;
            }
        }
        return min;
    }
}

File: app/beaglesightlibs/src/main/java/com/cross/beaglesightlibs/bowconfigs/PositionCalculator.kt
package com.cross.beaglesightlibs.bowconfigs

import java.text.DecimalFormat

abstract class PositionCalculator {
    var positions: List<PositionPair>? = null

    abstract fun calcPosition(distance: Float): Float

    //TODO: Horrifically slow and inaccurate. Replace with optimised version?
    fun getMaxPosition(distStart: Float, distEnd: Float): Float {
        var max = Float.MIN_VALUE
        var i = distStart
        while (i < distEnd) {
            val `val` = calcPosition(i)
            if (`val` > max) {
                max = `val`
            }
            i++
        }
        return max
    }

    //TODO: Horrifically slow and inaccurate. Replace with optimised version?
    fun getMinPosition(distStart: Float, distEnd: Float): Float {
        var min = Float.MAX_VALUE
        var i = distStart
        while (i < distEnd) {
            val `val` = calcPosition(i)
            if (`val` < min) {
                min = `val`
            }
            i++
        }
        return min
    }

    companion object {
        private val zero = DecimalFormat("#")
        private val one = DecimalFormat("#.#")
        private val two = DecimalFormat("#.##")
        fun getDisplayValue(value: Float, numPlaces: Int): String {
            return when (numPlaces) {
                0 -> zero.format(value)
                1 -> one.format(value)
                2 -> two.format(value)
                else -> two.format(value)
            }
        }
    }
}